src/main/java/at/ac/tuwien/kr/alpha/grounder/structure/AtomChoiceRelation.java

```diff
@@ -0,0 +1,31 @@
+package at.ac.tuwien.kr.alpha.grounder.structure;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Stores and provides relationships between ordinary atoms and those that represent choice points.
+ *
+ * Copyright (c) 2019, the Alpha Team.
+ */
+public class AtomChoiceRelation {
+	private final ArrayList<ArrayList<Integer>> atomToChoiceAtoms = new ArrayList<>();
```

Comment:
I like to avoid `List` for any variable that is performance-significant. `ArrayList` more explicitly states that no slow implementation should be used. Anyway, I'm changing it to an array (which also removes some of the weird-looking add-loops).

Semantically, it can be seen as a set of atoms for each atom. For performance reasons (and because we know the grounder does not ground the same rule twice), a list is sufficient here.

Reply:
How can you be so sure that `ArrayList` actually is the best-performing implementation in this case? Elements are appended always to the end of the list, couldn´t it be that a `LinkedList`, for example, benefits from this? Performance also depends on the number of inserts and reads, afaik.

Regarding set semantics:
For rules that contain the head in the body, e.g., `x :- not x.`, the list will contain the body twice. I think we can disregard this case.

Reply:
Performance depends on usage, yes. For this case, we have to distinguish between the inner and outer collection.

The outer `ArrayList` has for each known atom one entry and is accessed in a random fashion, no entries are deleted and nothing is inserted (addition happens when `growForMaxAtomId` is executed). Here we can be pretty sure that `LinkedList` has way worse performance than `ArrayList`, because the random access in `LinkedList` is terrible O(n) while adding at the end of an `ArrayList` is on average as efficient as adding to the `LinkedList` namely O(1), but of course when the array has to grow, it is O(n), but this happens only log m times for m bein the final size of the array.

For the inner `ArrayList` there is only sequential processing of the list (outside of `AtomChoiceRelation`) and we only add entries at the end. In principle, `LinkedList` and `ArrayList` could perform similarly for this one. I guess, one might have to benchmark that to see a difference. I still would not use a `LinkedList` though, since it likely is less memory-efficient.

`LinkedList` really has a limited use, namely cases where you know that access is mostly sequential and many insertions/deletion at random points inside the list will be done.